core/src/main/java/org/infinispan/util/ByteString.java

```diff
@@ -0,0 +1,93 @@
+package org.infinispan.util;
+
+import org.infinispan.commons.marshall.AdvancedExternalizer;
+import org.infinispan.marshall.core.Ids;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * A simple class which encapsulates a byte[] representation of a String using a predefined encoding (currently UTF-8).
+ * This avoids repeated invocation of the expensive {@link ObjectOutput#writeUTF(String)} on marshalling
+ *
+ * @author Tristan Tarrant
+ * @since 9.0
+ */
+public class ByteString {
+   public static final Charset CHARSET = Charset.forName("UTF-8");
+   private static final ByteString EMPTY = new ByteString(new byte[0]);
+   private final byte[] b;
+   private final transient int hash;
+
+   ByteString(byte[] b) {
+      this.b = b;
+      this.hash = Arrays.hashCode(b);
+   }
+
+   public static ByteString fromString(String s) {
+      if (s.length() == 0)
+         return EMPTY;
+      else
+         return new ByteString(s.getBytes(CHARSET));
+   }
+
+   public static ByteString emptyString() {
+      return EMPTY;
+   }
+
+   @Override
+   public int hashCode() {
+      return hash;
+   }
+
+   @Override
+   public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      ByteString that = (ByteString) o;
+      return Arrays.equals(b, that.b);
+   }
+
+   @Override
+   public String toString() {
+      return new String(b, CHARSET);
```

Comment:
AFAICT this is only used in `LogFactory.pushNDC(ByteString, boolean)`, and that is only used for `StateRequestCommand`, and `StateResponseCommand`, so it shouldn't be a problem. 

And even those usages could be eliminated by calling `LogFactory.pushNDC()` from the inbound invocation handler for all commands, instead of `InvocationContextInterceptor` for some and `perform()` for others.